File: corevo_fe/Hit_Tech/lib/service/auth_service.dart
import 'package:dio/dio.dart';
import 'package:hit_tech/core/constants/api_endpoint.dart';
import 'package:hit_tech/model/request/auth/forgot_password_request.dart';
import 'package:hit_tech/model/request/auth/oauth2_google_request.dart';
import 'package:hit_tech/model/response/auth/forgot_password_response.dart';
import 'package:hit_tech/model/response/auth/login_response.dart';
import 'package:hit_tech/model/response/auth/reset_password_response.dart';
import 'package:hit_tech/model/response/default_response.dart';
import 'package:hit_tech/service/shared_preferences.dart';
import 'package:hit_tech/utils/dio_client.dart';
import 'package:http/http.dart' as http;
import 'dart:convert';

import '../model/request/auth/login_request.dart';
import '../model/request/auth/register_request.dart';
import '../model/request/auth/reset_password_request.dart';
import '../model/request/auth/verify_otp_request.dart';
import '../model/response/auth/register_response.dart';
import '../model/response/auth/verify_opt_response.dart';

class AuthService {
  static Future<LoginResponse> login(LoginRequest request) async {
    try {
      final response = await DioClient.dio.post(
        ApiEndpoint.login,
        data: request.toJson(),
        options: Options(
          contentType: Headers.jsonContentType,
          sendTimeout: Duration(seconds: 5),
          receiveTimeout: Duration(seconds: 5),
        ),
      );

      return LoginResponse.fromJson(response.data);
    } on DioException catch (e) {
      print('DIO ERROR: ${e.message}');
      print('RESPONSE: ${e.response?.data}');
      rethrow;
    } catch (e, stack) {
      print('ERROR: $e');
      print('STACKTRACE: $stack');
      rethrow;
    }
  }

  static Future<LoginResponse> loginWithGoogle(
    Oauth2GoogleRequest request,
  ) async {
    try {
      final response = await DioClient.dio.post(
        ApiEndpoint.loginWithGoogle,
        data: request.toJson(),
        options: Options(
          contentType: Headers.jsonContentType,
          sendTimeout: Duration(seconds: 5),
          receiveTimeout: Duration(seconds: 5),
        ),
      );

      return LoginResponse.fromJson(response.data);
    } on DioException catch (e) {
      print('DIO ERROR: ${e.message}');
      print('RESPONSE: ${e.response?.data}');
      rethrow;
    } catch (e, stack) {
      print('ERROR: $e');
      print('STACKTRACE: $stack');
      rethrow;
    }
  }

  static Future<RegisterResponse> register(RegisterRequest request) async {
    try {
      final response = await DioClient.dio.post(
        ApiEndpoint.register,
        data: request.toJson(),
        options: Options(
          contentType: Headers.jsonContentType,
          sendTimeout: Duration(seconds: 5),
          receiveTimeout: Duration(seconds: 5),
        ),
      );

      return RegisterResponse.fromJson(response.data);
    } on DioException catch (e) {
      print('DIO ERROR: ${e.message}');
      print('RESPONSE: ${e.response?.data}');
      rethrow;
    } catch (e, stack) {
      print('ERROR: $e');
      print('STACKTRACE: $stack');
      rethrow;
    }
  }

  static Future<VerifyOtpResponse> verifyOtpToRegister(
    VerifyOtpRequest request,
  ) async {
    try {
      final response = await DioClient.dio.post(
        ApiEndpoint.verifyOtp,
        data: request.toJson(),
        options: Options(
          contentType: Headers.jsonContentType,
          sendTimeout: Duration(seconds: 5),
          receiveTimeout: Duration(seconds: 5),
        ),
      );

      return VerifyOtpResponse.fromJson(response.data);
    } on DioException catch (e) {
      print('DIO ERROR: ${e.message}');
      print('RESPONSE: ${e.response?.data}');
      rethrow;
    } catch (e, stack) {
      print('ERROR: $e');
      print('STACKTRACE: $stack');
      rethrow;
    }
  }

  static Future<ForgotPasswordResponse> sendEmailToForgotPassword(
    ForgotPasswordRequest request,
  ) async {
    try {
      final response = await DioClient.dio.post(
        ApiEndpoint.sendEmailToForgotPassword,
        data: request.toJson(),
        options: Options(
          contentType: Headers.jsonContentType,
          sendTimeout: Duration(seconds: 5),
          receiveTimeout: Duration(seconds: 5),
        ),
      );

      return ForgotPasswordResponse.fromJson(response.data);
    } on DioException catch (e) {
      print('DIO ERROR: ${e.message}');
      print('RESPONSE: ${e.response?.data}');
      rethrow;
    } catch (e, stack) {
      print('ERROR: $e');
      print('STACKTRACE: $stack');
      rethrow;
    }
  }

  static Future<VerifyOtpResponse> verifyOtpToResetPassword(
    VerifyOtpRequest request,
  ) async {
    try {
      final response = await DioClient.dio.post(
        ApiEndpoint.verifyOptResetPassword,
        data: request.toJson(),
        options: Options(
          contentType: Headers.jsonContentType,
          sendTimeout: Duration(seconds: 5),
          receiveTimeout: Duration(seconds: 5),
        ),
      );

      return VerifyOtpResponse.fromJson(response.data);
    } on DioException catch (e) {
      print('DIO ERROR: ${e.message}');
      print('RESPONSE: ${e.response?.data}');
      rethrow;
    } catch (e, stack) {
      print('ERROR: $e');
      print('STACKTRACE: $stack');
      rethrow;
    }
  }

  static Future<ResetPasswordResponse> resetPassword(
    ResetPasswordRequest request,
  ) async {
    try {
      final response = await DioClient.dio.post(
        ApiEndpoint.resetPassword,
        data: request.toJson(),
        options: Options(
          contentType: Headers.jsonContentType,
          sendTimeout: Duration(seconds: 5),
          receiveTimeout: Duration(seconds: 5),
        ),
      );

      return ResetPasswordResponse.fromJson(response.data);
    } on DioException catch (e) {
      print('DIO ERROR: ${e.message}');
      print('RESPONSE: ${e.response?.data}');
      rethrow;
    } catch (e, stack) {
      print('ERROR: $e');
      print('STACKTRACE: $stack');
      rethrow;
    }
  }

  static Future<DefaultResponse> logout(String request) async {
    try {
      final response = await DioClient.dio.post(
        ApiEndpoint.logout,
        data: {'token': request},
        options: Options(
          contentType: Headers.jsonContentType,
          sendTimeout: Duration(seconds: 5),
          receiveTimeout: Duration(seconds: 5),
        ),
      );

      return DefaultResponse.fromJson(response.data);
    } on DioException catch (e) {
      print('DIO ERROR: ${e.message}');
      print('RESPONSE: ${e.response?.data}');
      rethrow;
    } catch (e, stack) {
      print('ERROR: $e');
      print('STACKTRACE: $stack');
      rethrow;
    }
  }

  static Future<String?> refreshAccessToken() async {
    final refreshToken = await SharedPreferencesService.getRefreshToken();
    if (refreshToken == null) return null;

    final response = await http.post(
      Uri.parse(ApiEndpoint.refreshToken),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({'refreshToken': refreshToken}),
    );

    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      final newAccessToken = data['data']['accessToken'];
      if (newAccessToken != null) {
        await SharedPreferencesService.saveAccessToken(newAccessToken);
        return newAccessToken;
      } else {
        await SharedPreferencesService.clearAll();
        return null;
      }
    } else {
      await SharedPreferencesService.clearAll();
      return null;
    }
  }

  static Future<DefaultResponse> getPrivacy() async {
    try {
      final response = await DioClient.dio.get(
        ApiEndpoint.getPolicyPrivacy,
        options: Options(
          contentType: Headers.jsonContentType,
          sendTimeout: Duration(seconds: 5),
          receiveTimeout: Duration(seconds: 5),
        ),
      );

      return DefaultResponse.fromJson(response.data);
    } on DioException catch (e) {
      print('DIO ERROR: ${e.message}');
      print('RESPONSE: ${e.response?.data}');
      rethrow;
    } catch (e, stack) {
      print('ERROR: $e');
      print('STACKTRACE: $stack');
      rethrow;
    }
  }

  static Future<DefaultResponse> getTerms() async {
    try {
      final response = await DioClient.dio.get(
        ApiEndpoint.getPolicyTerms,
        options: Options(
          contentType: Headers.jsonContentType,
          sendTimeout: Duration(seconds: 5),
          receiveTimeout: Duration(seconds: 5),
        ),
      );

      return DefaultResponse.fromJson(response.data);
    } on DioException catch (e) {
      print('DIO ERROR: ${e.message}');
      print('RESPONSE: ${e.response?.data}');
      rethrow;
    } catch (e, stack) {
      print('ERROR: $e');
      print('STACKTRACE: $stack');
      rethrow;
    }
  }

  static Future<DefaultResponse> sendEmailToRecoveryAccount(
      String email,
      ) async {
    try {
      final response = await DioClient.dio.post(
        ApiEndpoint.accountRecover,
        data: {
          "email": email
        },
        options: Options(
          contentType: Headers.jsonContentType,
          sendTimeout: Duration(seconds: 5),
          receiveTimeout: Duration(seconds: 5),
        ),
      );

      return DefaultResponse.fromJson(response.data);
    } on DioException catch (e) {
      print('DIO ERROR: ${e.message}');
      print('RESPONSE: ${e.response?.data}');
      rethrow;
    } catch (e, stack) {
      print('ERROR: $e');
      print('STACKTRACE: $stack');
      rethrow;
    }
  }

  static Future<VerifyOtpResponse> verifyOtpToRecover(
      VerifyOtpRequest request,
      ) async {
    try {
      final response = await DioClient.dio.post(
        ApiEndpoint.recoverAccount,
        data: request.toJson(),
        options: Options(
          contentType: Headers.jsonContentType,
          sendTimeout: Duration(seconds: 5),
          receiveTimeout: Duration(seconds: 5),
        ),
      );

      return VerifyOtpResponse.fromJson(response.data);
    } on DioException catch (e) {
      print('DIO ERROR: ${e.message}');
      print('RESPONSE: ${e.response?.data}');
      rethrow;
    } catch (e, stack) {
      print('ERROR: $e');
      print('STACKTRACE: $stack');
      rethrow;
    }
  }

}

File: corevo_fe/Hit_Tech_React_Native/my-app/services/authService.ts
import apiClient from '@/services/api/client';
import { ApiEndpoints } from '@/constants/ApiEndpoints';
import {
    LoginRequest,
    LoginResponse,
    RegisterRequest,
    RegisterResponse,
    VerifyOtpRequest,
    VerifyOtpResponse,
    ForgotPasswordRequest,
    ForgotPasswordResponse,
    ResetPasswordRequest,
    ResetPasswordResponse,
    OAuth2GoogleRequest,
    DefaultResponse,
} from '@/types/auth';
import { ApiResponse } from '@/types/common';
import { saveAccessToken, saveRefreshToken, getRefreshToken, clearTokens } from '@/services/storage';

/**
 * Authentication Service
 * All authentication-related API calls
 */

/**
 * Login with username and password
 */
export const login = async (request: LoginRequest): Promise<ApiResponse<LoginResponse>> => {
    const response = await apiClient.post<ApiResponse<LoginResponse>>(
        ApiEndpoints.auth.login,
        request
    );
    return response.data;
};

/**
 * Login with Google OAuth2
 */
export const loginWithGoogle = async (
    request: OAuth2GoogleRequest
): Promise<ApiResponse<LoginResponse>> => {
    const response = await apiClient.post<ApiResponse<LoginResponse>>(
        ApiEndpoints.auth.loginWithGoogle,
        request
    );
    return response.data;
};

/**
 * Register new user
 */
export const register = async (request: RegisterRequest): Promise<ApiResponse<RegisterResponse>> => {
    const response = await apiClient.post<ApiResponse<RegisterResponse>>(
        ApiEndpoints.auth.register,
        request
    );
    return response.data;
};

/**
 * Verify OTP for registration
 */
export const verifyOtpToRegister = async (
    request: VerifyOtpRequest
): Promise<ApiResponse<VerifyOtpResponse>> => {
    const response = await apiClient.post<ApiResponse<VerifyOtpResponse>>(
        ApiEndpoints.auth.verifyOtp,
        request
    );
    return response.data;
};

/**
 * Send email to forgot password
 */
export const sendEmailToForgotPassword = async (
    request: ForgotPasswordRequest
): Promise<ApiResponse<ForgotPasswordResponse>> => {
    const response = await apiClient.post<ApiResponse<ForgotPasswordResponse>>(
        ApiEndpoints.auth.forgotPassword,
        request
    );
    return response.data;
};

/**
 * Verify OTP for reset password
 */
export const verifyOtpToResetPassword = async (
    request: VerifyOtpRequest
): Promise<ApiResponse<VerifyOtpResponse>> => {
    const response = await apiClient.post<ApiResponse<VerifyOtpResponse>>(
        ApiEndpoints.auth.verifyOtpResetPassword,
        request
    );
    return response.data;
};

/**
 * Reset password
 */
export const resetPassword = async (
    request: ResetPasswordRequest
): Promise<ApiResponse<ResetPasswordResponse>> => {
    const response = await apiClient.post<ApiResponse<ResetPasswordResponse>>(
        ApiEndpoints.auth.resetPassword,
        request
    );
    return response.data;
};

/**
 * Logout user
 */
export const logout = async (token: string): Promise<ApiResponse<DefaultResponse>> => {
    const response = await apiClient.post<ApiResponse<DefaultResponse>>(
        ApiEndpoints.auth.logout,
        { token }
    );
    return response.data;
};

/**
 * Refresh access token
 */
export const refreshAccessToken = async (): Promise<string | null> => {
    const refreshToken = await getRefreshToken();

    if (!refreshToken) {
        return null;
    }

    try {
        const response = await apiClient.post(
            ApiEndpoints.auth.refreshToken,
            { refreshToken }
        );

        const newAccessToken = response.data.data.accessToken;

        if (newAccessToken) {
            await saveAccessToken(newAccessToken);
            return newAccessToken;
        } else {
            await clearTokens();
            return null;
        }
    } catch (error) {
        await clearTokens();
        return null;
    }
};

/**
 * Get privacy policy
 */
export const getPrivacy = async (): Promise<ApiResponse<DefaultResponse>> => {
    const response = await apiClient.get<ApiResponse<DefaultResponse>>(
        ApiEndpoints.policy.getPrivacy
    );
    return response.data;
};

/**
 * Get terms of service
 */
export const getTerms = async (): Promise<ApiResponse<DefaultResponse>> => {
    const response = await apiClient.get<ApiResponse<DefaultResponse>>(
        ApiEndpoints.policy.getTerms
    );
    return response.data;
};

/**
 * Send email to recover account
 */
export const sendEmailToRecoveryAccount = async (
    email: string
): Promise<ApiResponse<DefaultResponse>> => {
    const response = await apiClient.post<ApiResponse<DefaultResponse>>(
        ApiEndpoints.auth.accountRecovery,
        { email }
    );
    return response.data;
};

/**
 * Verify OTP to recover account
 */
export const verifyOtpToRecover = async (
    request: VerifyOtpRequest
): Promise<ApiResponse<VerifyOtpResponse>> => {
    const response = await apiClient.post<ApiResponse<VerifyOtpResponse>>(
        ApiEndpoints.auth.recoverAccount,
        request
    );
    return response.data;
};
